src/main/java/org/broadinstitute/hellbender/tools/walkers/haplotypecaller/PartiallyDeterminedHaplotypeComputationEngine.java

```diff
@@ -806,86 +788,60 @@ public void populateBitset(List<List<Event>> disallowedEvents) {
          * @param disallowSubsets
          * @return
          */
-        public List<List<Tuple<Event,Boolean>>> getVariantGroupsForEvent(final List<Event> allEventsHere, final int determinedAlleleIndex, final boolean disallowSubsets) {
-            // If we are dealing with an external to this list event
-            int eventMask = 0;
-            int maskValues = 0;
-            for(int i = 0; i < allEventsHere.size(); i++) {
-                if (eventSet.contains(allEventsHere.get(i))) {
-                    int index = eventsInBitmapOrder.indexOf(allEventsHere.get(i));
-                    eventMask = eventMask | (1 << index);
-                    maskValues = maskValues | ((i == determinedAlleleIndex ? 1 : 0) << index);
-                }
-            }
+        public List<Set<Event>> setsForBranching(final List<Event> locusEvents, final Set<Event> determinedEvents, final boolean disallowSubsets) {
+            final SmallBitSet locusOverlapSet = overlapSet(locusEvents);
+            final SmallBitSet determinedOverlapSet = overlapSet(determinedEvents);
+
             // Special case (if we are determining bases outside of this mutex cluster we can reuse the work from previous iterations)
-            if (eventMask == 0 && cachedEventLists != null) {
-                return cachedEventLists;
+            if (locusOverlapSet.isEmpty() && cachedEventSets != null) {
+                return cachedEventSets;
             }
 
-            List<Integer> ints = new ArrayList<>();
+            final List<SmallBitSet> allowedAndDetermined = new ArrayList<>();
             // Iterate from the BACK of the list (i.e. ~supersets -> subsets)
             // NOTE: we skip over 0 here since that corresponds to ref-only events, handle those externally to this code
-            outerLoop:
-            for (int i = allowedEvents.length(); i > 0; i--) {
-                // If the event is allowed AND if we are looking for a particular event to be present or absent.
-                if (allowedEvents.get(i) && (eventMask == 0 || ((i & eventMask) == maskValues))) {
+            for (final SmallBitSet subset = SmallBitSet.fullSet(eventsInOrder.size()); !subset.isEmpty(); subset.decrement()) {
```

Comment:
re-reading this eventhough I wrote it using i-- as a set it occurs to me just how confusing this behavior is... make it a bit clearer in the comments here that this makes sense.

Reply:
done